### ISSUE TITLE
Adding a `TextItem` to a `PlotItem` with no other data breaks auto-scaling

### ISSUE THREAD

**Issue by redruin1** [pyqtgraph/pyqtgraph#2642]
### Short description
I'm attempting to add text labels to a standard `PlotItem` widget. In my production code however, I neglected to add any data before I added the text (I was adding ROIs as "data"), and it resulted in both the x and y scales to trend towards 0 very quickly when auto-scaling is on.

### MinRe Example
```python
import pyqtgraph as pg

if __name__ == "__main__":
    pg.mkQApp()

    pw = pg.PlotWidget()
    pw.show()

    # Having a plot with no data works fine, until you add the TextItem:
    text_item = pg.TextItem("Some text")
    pw.addItem(text_item)

    # Including either (or both) of the following lines produce expected behavior:
    #pw.plot([1, 2, 3], [4, 5, 6])
    #pw.getPlotItem().getViewBox().disableAutoRange()

    pg.exec()
```

### Expected behavior
Either the standard graph display centered around the text "Some text" located at the origin, or the standard graph centered around the range [0, 1] (which when panned to the origin reveals the aforementioned text).

### Real behavior
For the first few seconds of the display, the scale factors are seen changing rapidly every frame; upon close inspection, they can be seen shrinking until they approach 0:

![image](https://user-images.githubusercontent.com/25079873/223302091-e8eb15c5-491b-4709-b152-982506d119d4.png)

Eventually the graph markers break completely as the x and y scales sit precariously above 0, as revealed by the context menu:

![image](https://user-images.githubusercontent.com/25079873/223303232-11256e8d-9cc6-412f-bd45-13e72ddb29fd.png)

Switching the axis modes from Auto to Manual and setting an appropriate range provides the expected outcome:

![image](https://user-images.githubusercontent.com/25079873/223303515-5b0a40aa-9c21-4d96-bdf2-677a45190808.png)

### Tested environment

 * PyQtGraph version: `0.13.2`
 * Qt Python binding: `'PyQt5 5.15.9 Qt 5.15.2'`
 * Python version: `3.10.0`
 * NumPy version: `1.24.2`
 * Operating system: Windows 10
 * Installation method: `pip install pyqtgraph`

**Comment by j9ac9k** [comment]
Thanks for the report @redruin1 !

I'm a _little_ on the fence about this being a bug.  TextItem is text whose size is independent of the view scale, anchored at a point.  When you add a TextItem to a view, you're adding a point to the view and have some text in another coordinate system get mapped to that point.

What's happening here is that as far as the view is concerned, the only thing present is the point (for which the TextItem is anchored to).  So when you autoscale, you're getting what is likely the smallest possible range for pyqtgraph to display as you're trying to make the view fit a single point.

Going to your expected behavior:

> Either the standard graph display centered around the text "Some text" located at the origin, or the standard graph centered around the range [0, 1] (which when panned to the origin reveals the aforementioned text).

So the graph won't center on the text (it will center on the anchor point), but you can still get _extremely_ small view ranges, but I think when adding a TextItem (or something with just a point) then maybe preserving the view range to [0, 1] that might be reasonable.  Figuring out how to do that w/o adding significantly more complexity to addItem might be tricky tho.

**Comment by redruin1** [comment]
Thanks for the prompt reply @j9ac9k, best-fitting to the smallest range gives a good explanation for the behavior I'm seeing. However, after hearing your explanation I wondered what would happen to autoscaling if you tried to plot a single point as data (ie, on a graph):

```python
pg.mkQApp()

pw = pg.PlotWidget()
pw.plot([0], [0])

pw.show()
pg.exec()
```

This behavior seems supported and well defined: the range [0, 1] centered around the data point:

![image](https://user-images.githubusercontent.com/25079873/223743058-c8fe83ba-e0c6-473b-ba14-3c4eaf8cf439.png)

Could the code for this be tweaked somewhat to handle TextItems? I don't think you'd have to take the text's dimension into account, but something like the above (at least to me) makes much more sense from a user-experience standpoint. 

Furthermore, it seems that TextItems are already treated like data points internally in some capacity, because if you plot them alongside other data the view encompasses the text alongside:

```python
pg.mkQApp()

pw = pg.PlotWidget()
pw.plot([1, 2, 3], [4, 5, 6]) # Some data, not at the origin

text_item = pg.TextItem("Some text")
pw.addItem(text_item) # Text at the origin

pw.show()
pg.exec()
```

![image](https://user-images.githubusercontent.com/25079873/223745558-c035335c-15b2-428a-a586-2bfd4268f764.png)

Regardless, I'll defer to the decision of the maintainers. Even something like a note in the documentation would be a great help, since for a while I was scratching my head as to why this bug was happening, as it's cause is not self-evident.

**Comment by j9ac9k** [comment]
> Could the code for this be tweaked somewhat to handle TextItems? I don't think you'd have to take the text's dimension into account, but something like the above (at least to me) makes much more sense from a user-experience standpoint.

I agree with this.  While I think an argument could be made that this isn't a bug and is "working as intended", I don't think that's a particularly great argument, and we can (and should) do better.

I haven't examined it carefully but looks like we have some precedence for treating different GraphicsItem's differently.

https://github.com/pyqtgraph/pyqtgraph/blob/3d3344d335d936209f65bf521cf24ff68f758302/pyqtgraph/graphicsItems/PlotItem/PlotItem.py#L501

I took a look at [TextItem.py](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/graphicsItems/TextItem.py) and I'm not seeing any obvious place here to modify the code.

Another thing you can try and see if you get desired behavior is replicate with [TargetItem](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/graphicsItems/TargetItem.py) that has similar behavior as the TextItem, if that works as intended, probably not much effort needed to adopt to TextItem.

**Comment by redruin1** [comment]
After skimming through the code, it seems the reason why a single data point graph works is because the bounding area of a single data point is always a zero-area rectangle. pyqtgraph can identify this case, and properly defaults to the [0, 1] range centered around the point. In a similar vein, `TargetItem` has the `dataBounds()` function which returns `[None, None]`, which is also correctly handled.

With `TextItem`s however, `dataBounds()` is not defined, so it gets their `boundingRect()` instead. This would be fine, but I believe that it doesn't take into account the visual size of the text and instead tries to scale the text to fill as much of the plot as possible. But since the visual size of these elements are supposed to remain constant as a fraction of the ViewBox's area, the text box is scaled down to maintain it's visual size, causing the autorange to scale the view to maximize the size of the text, causing the text box to shrink again, etc.

The top of `ViewBox.updateAutoRange()` implies that it's supposed to guard against these kinds of recursive resizing problem, but for some reason it seems to not be catching this particular case:

```python
def updateAutoRange(self):
    ## Break recursive loops when auto-ranging.
    ## This is needed because some items change their size in response
    ## to a view change.
    if self._updatingRange:
        return
    self._updatingRange = True
    try:
        # ...
```

I'm not familiar enough with the code yet to diagnose an exact point of failure though.

Intuition would suggest a special case handling when there's only one TextItem, but that's actually not sufficient, since it can also happen with multiple as long as they're along the same axis:

```python
text_item = pg.TextItem("Some text")
pw.addItem(text_item)

other_text_item = pg.TextItem("Some other text")
other_text_item.setPos(10, 0)
pw.addItem(other_text_item)
```

This causes the runaway scaling, but only across the y-axis. Placing `other_text_item` at (10, 10) causes no runaway scaling at all.

A very quick and very dirty fix I found was to implement the `dataBounds()` function for `TextItem` to instead grab the text's anchor point instead of it's bounding rect:

```python
def dataBounds(self, ax, frac=1.0, orthoRange=None):
        """
        TEST: what if we treat the text box as a single point instead of a bounding rectangle?
        """
        result = [None, None]
        pos_on_axis = self.anchor[ax]
        
        if orthoRange:
            # If range is specified, make sure the anchor is within the range
            # before we return it
            range_min, range_max = orthoRange[0], orthoRange[1]
            if range_min <= pos_on_axis <= range_max:
                result = [pos_on_axis, pos_on_axis]
        else:
            # Otherwise, just return the solitary point
            result = [pos_on_axis, pos_on_axis]

        return result
```

Of course, this wouldn't _guarantee_ that the text itself would be visible (which very well may be another specification failure), but it produces the same range result as a graph where each text item's anchor is a data point.

TLDR: Ideally `boundingRect()` would take into account the font size of the text rendered so that this runaway scaling doesn't happen, or there should be a fix to the anti-recursion code to head the runaway scaling off at the pass. Alternatively, one could also just ignore the box's area and just get the anchor point, though it sacrifices some visual polish. Thoughts?

**Comment by j9ac9k** [comment]
I'm reluctant on tweaking boundingRect() for this purpose; as TextItem maintains the fixed text size _regardless_ of the view scale, when you zoom in some way, the boundingRect() will be recalculated, triggering an autoscale recalculation, triggering a new scale operation, triggering a new boundingRect calculation.

In terms of why the ViewBox recursion isn't catching it, I think that catch is meant to capture recursion that is triggered by code internal to ViewBox; furthermore "here be dragons" in `ViewBox.py` 😆 

I think applying a dataBounds() to TextItem would be the right way to fix this.  I would have to tinker with a PR to try and get a case where the text is off the screen when it shouldn't be.

I think the dataBounds method you posted _should_ work (again, just eye-balling w/o doing any testing).  If you want to submit a PR with that change, I'll try and tinker with it.  

Thank you for investigating this issue.

**Comment by redruin1** [comment]
@j9ac9k, created a PR. Enclosed are the upsides and downsides that I've found, though let me know if I've caused other issues elsewhere. Modification is minimal though and tests seem to pass fine.

**Comment by Kanonck** [comment]
Hi, I encountered this thread because I have the same issue. I have the newest version of pyqtgraph, so the PR with the fix is part of my code.
I have a textitem, that I want to center in the middle and I constantly update its position, to ensure it remains in the centre. I have a plot with several signals, but If I plot with some signals that have no data, then my y-axis starts "running" towards 0 until my application breaks.
Do I need to manually call dataBounds()?

**Comment by redruin1** [comment]
@Kanonck, can you make a minimum reproducible example? It's been a hot minute since I've taken a look at the internals, so I'm hoping a paste-able example will jog my memory.

**Comment by Kanonck** [comment]
@redruin1 sorry for the long silence, but here is an example. When I start it up, it immedeately comes into the auto-scaling issue and then breaks.

```
import sys
from PySide6.QtWidgets import QApplication, QMainWindow, QVBoxLayout, QWidget
import pyqtgraph as pg

class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()

        self.setWindowTitle("PySide6 and PyQtGraph Example")

        # Create a central widget and set the layout
        central_widget = QWidget()
        layout = QVBoxLayout(central_widget)
        self.setCentralWidget(central_widget)

        # Create a plot widget
        self.plot_widget = pg.PlotWidget()
        layout.addWidget(self.plot_widget)

        # Add a constant data line (y = 0)
        x = list(range(100))  # x values from 0 to 99
        y = [0] * 100  # y values are all 0
        self.plot_widget.plot(x, y, pen=pg.mkPen(color='b', width=2))

        # Add a text item to the plot
        text_item = pg.TextItem(text="Constant Line y=0", color='r', anchor=(0.5, 0.5))
        text_item.setPos(50, 0)  # Position the text item at (50, 0)
        self.plot_widget.addItem(text_item)

if __name__ == "__main__":
    app = QApplication(sys.argv)
    window = MainWindow()
    window.show()
    sys.exit(app.exec())

```

**Comment by Kanonck** [comment]
Any update on this?
